### ISSUE TITLE
Support fully-qualified names from wildcard imports across languages

### ISSUE THREAD

**Issue by ievans** [semgrep/semgrep#4839]
**Describe the bug**
Semgrep doesn't handle fully qualified names that are introduced via a wildcard import.

**To Reproduce**
Python example: https://semgrep.dev/s/RgQg
Java example: https://semgrep.dev/s/BGQx

**Expected behavior**
In the case where there is a single wildcard import, Semgrep should be able to infer that the unresolved method came from the wildcard. 
In the case where are are multiple wildcard imports, we could consider:
* embedding knowledge about the standard library
* performing more advanced name resolution
* conservative: assuming every unresolved name is present in every wildcard import and reporting there

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [x] P2: regular bug that should get fixed

**Environment**
Semgrep 0.85.

**Use case**
Increased rule coverage (avoid false negatives).

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-1006/support-fully-qualified-names-from-wildcard-imports-across-languages.

**Comment by aryx** [comment]
or D: leverage information from the pattern itself! if the pattern is new java.util.ArrayList(), then for sure such an entity exist and having import java.util.* and new ArrayList() is in the same file should be a match (without having to know about the content of java.util.*)

**Comment by ELLIOTTCABLE** [comment]
So, I started trying to take a look at this. It's only my second contribution, so that may have been ambitious, lol.

At first, it looks like I need to add an `ImportAll` case around here:

https://github.com/returntocorp/semgrep/blob/d9bd139e8bab8d76bc9e7d25707546b7f4f6ac9a/semgrep-core/src/naming/Naming_AST.ml#L547-L553

… then I realized that we'd need information about which modules/whatever are currently "open". It's a weird abuse of the mechanic, but I almost feel like I could achieve this by adding a new `scope`? Instead of a scope-of-bindings, it's a scope-of-openings?

https://github.com/returntocorp/semgrep/blob/d9bd139e8bab8d76bc9e7d25707546b7f4f6ac9a/semgrep-core/src/naming/Naming_AST.ml#L167-L169

Having added such a scope, and making sure `DottedName`s (right?) get added to it when an `ImportAll` is seen, then I *think* it's the `m_name` matcher that needs to get modified to (after a normal match fails) try removing any `open`ed components from scopes:

https://github.com/returntocorp/semgrep/blob/d9bd139e8bab8d76bc9e7d25707546b7f4f6ac9a/semgrep-core/src/matching/Generic_vs_generic.ml#L369-L374

That's all very sketched-out — and is about as far as I was able to get with my time today. In particular, it didn't look like anything from `Naming_AST` is used in there, except indirectly …

I also came across `m_id_info`, which looks relevant, but again, by that time, I'd run short on time.

https://github.com/returntocorp/semgrep/blob/d9bd139e8bab8d76bc9e7d25707546b7f4f6ac9a/semgrep-core/src/matching/Generic_vs_generic.ml#L527-L535

In particular, this comment.

https://github.com/returntocorp/semgrep/blob/d9bd139e8bab8d76bc9e7d25707546b7f4f6ac9a/semgrep-core/src/matching/Generic_vs_generic.ml#L550-L556

Hopefully, I'll get to return to this at some point — but 1. I hope someone with more familiarity with the codebase can tell me if I'm completely wasting my time (is the sketch above the correct approach, do you think?); and 2. if I am not able to return to it, I hope this can help someone else get started on a fix!

**Comment by aryx** [comment]
yes I think it's a quite complex change. I don't even know how to fix the issue. I'm not sure it's something we can fix in Naming_AST.ml. I was thinking more to fix it in generic_vs_generic.ml, to remember there the list of ImportAll currently in scope and depending on the pattern and the list of ImportAll, decide whether or not a match is possible.
I'll try to look at it next week.

**Comment by kurt-r2c** [comment]
adding a +1 to this - star-importing is not generally considered idiomatic Java due to namespace cluttering (and depending on the library may introduce vulnerable code or exploit primitives unnecessarily).
